### ISSUE TITLE
pywbem.cim_http.ConnectionError: Socket error: check_hostname requires server_hostname

### ISSUE THREAD

**Issue by skolsuper** [pywbem/pywbem#150]
Hey,

Firstly, so glad to see this package come back to life, and really like the direction you're going with the changes.  I inherited a project that was using a copypasted local version of this package with python3 fixes, and I'm now trying to migrate it back to using the upstream version.  However I'm running into an issue I can't figure out.

**Repro:**

```python
import pywbem

conn = pywbem.WBEMConnection('https://192.168.1.1', creds=('someuser', 'password'))

conn.EnumerateInstances('OMC_Fan')
```

**Expected result:**  Information about the hardware (the `"OMC_Fan"` above is just an example, I get the same result whatever string I pass to `EnumerateInstances`)

**Actual result:** The exception in the title.  Stepping through it with the debugger, it looks like the request is never made as a sanity check in the python `ssl` module fails.

Environment:  python 3.5 virtualenv on MacOSX El Capitan

Any help would be greatly appreciated.

**Comment by andy-maier** [comment]
Hello. Glad to hear you like the direction.
From what I could quickly gather, a SSL socket user (pywbem) has to pass in a "server_hostname" option somewhere. The SSL in Python was fixed in 2014 to always accept server_hostname even if the underlying OpenSSL does not support SNI yet. I have taken a look into the cim_http module of pywbem, and it does not pass any server_hostname option.

I have very little experience with dealing with SSL in Python 3, I'm afraid, so I could use your help:

-> If your locally patched version works for this case, would you mind pasting its cim_http.py file into a comment of this issue?

**Comment by skolsuper** [comment]
Good news, after asking around, I found the history!  Somebody that actually understands git was the one that made the changes, and documented them.  I think [this](https://github.com/cloud-hm/pywbem3/commit/ca5c726daba4258b76fab35243b3d7e01619a322) is the relevant commit, but there are only a few commits in that repo, so hopefully the fix won't be difficult to find.  I am also looking but it's the end of the day in my timezone soon so I will resume tomorrow.  If I fix it I will submit a PR.

**Comment by KSchopmeyer** [comment]
Assigned to Karl to see if he can set up test.

**Comment by andy-maier** [comment]
Discussed today with Karl. Karl will try to investigate this using OP.

**Comment by skolsuper** [comment]
Possibly relevant, I happened across this in an unrelated repo: https://github.com/rethinkdb/rethinkdb/issues/5470

They're using a backport for `match_hostname` for python < 3.4, so maybe their code plus backport could be used by pywbem too.  The module is available [here](https://github.com/rethinkdb/rethinkdb/tree/next/drivers/python/rethinkdb/backports/ssl_match_hostname) with LICENSE and README

**Comment by andy-maier** [comment]
James,
would you mind trying out the change from PR #179 (branch socket_err_check_hostname) to see whether it addresses the problem you saw?

**Comment by skolsuper** [comment]
I will test it today and get back to you, thanks

**Comment by skolsuper** [comment]
I now get this error:

```
Traceback (most recent call last):
  File "/venvs/my-project/src/pywbem/pywbem/cim_http.py", line 613, in wbem_request
    client.endheaders()
  File "/usr/lib/python3.5/http/client.py", line 1079, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python3.5/http/client.py", line 911, in _send_output
    self.send(msg)
  File "/venvs/my-project/src/pywbem/pywbem/cim_http.py", line 354, in send
    self.connect()
  File "/venvs/my-project/src/pywbem/pywbem/cim_http.py", line 462, in connect
    ret = self.sock.connect(self.host, self.port)
TypeError: connect() takes 2 positional arguments but 3 were given

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "script_debug_vm.py", line 10, in <module>
    ret = conn.EnumerateInstances('OMC_Fan')
  File "/venvs/my-project/src/pywbem/pywbem/cim_operations.py", line 1114, in EnumerateInstances
    **params)
  File "/venvs/my-project/src/pywbem/pywbem/cim_operations.py", line 658, in imethodcall
    timeout=self.timeout)
  File "/venvs/my-project/src/pywbem/pywbem/cim_http.py", line 618, in wbem_request
    raise ConnectionError("Socket error: %s" % exc)
pywbem.cim_http.ConnectionError: Socket error: connect() takes 2 positional arguments but 3 were given
```

I feel like that is an improvement though, looks easier to debug.

**Comment by skolsuper** [comment]
Apologies, I had this hanging around at the top of my file:
```python
import ssl

ssl.create_default_context = ssl._create_unverified_context
```

as a relic from earlier attempts to get this to work.  Removing it results in the same `pywbem.cim_http.ConnectionError: Socket error: check_hostname requires server_hostname` error from before.

**Comment by KSchopmeyer** [comment]
Sorry to take so long with this one but while the problem is simple, the code documentation and the complexityof our code leaves something to be desired.

**Comment by skolsuper** [comment]
No problem at all, having looked at the source code in `cim_http`, I completely understand :).  Thanks for sharing your progress.

It wasn't around when this library was written, but I wonder if in 2016 we could be using the `requests` package for the low-level http drudgery.  It adds a dependency, but it might be the most commonly depended on 3rd party package in all of Python so in my opinion it's probably worth the cost.

That is one way we could push complexity "up".  I have submitted a pull request #224 with another approach, (optionally) pushing complexity "down" to the client by allowing them to inject their own `wbem_request` function.  When I have my own working and tested `Requester` class (probably `RequestsRequester` 😆) I will make another PR.

**Comment by andy-maier** [comment]
I agree that it would be worthwhile to use the `requests` package. In fact, Ross Peoples has come up with a new implementation of pywbem that uses that, and I'd like to integrate that.

**Comment by andy-maier** [comment]
Moving to v0.8.4.

**Comment by KSchopmeyer** [comment]
Yes we definitely want to consider moving to a package such as requests but we have to be very careful in a couple of areas, specifically security and implementation of things such as specific security requirements, implementation of mutual authentication, etc. to be sure we don't hurt outselves.

**Comment by KSchopmeyer** [comment]
We believe that we have resolved this issue with pr 179 and are creating a backport of 179 for the 0.8.4 release of pywbem

**Comment by andy-maier** [comment]
PR #179 (merged) addresses that for 0.9, and PR #298 (not yet merged) addresses that for 0.8.
Once PR #298 has been verified and merged, this issue can be closed.

**Comment by andy-maier** [comment]
PR #298 has been merged. Closing this issue.
